prefect_kubernetes/worker.py

```diff
@@ -0,0 +1,671 @@
+        """
+        try:
+            yield kubernetes.client.CoreV1Api(api_client=client)
+        finally:
+            client.rest_client.pool_manager.clear()
+
+    def _get_cluster_uid(self, client: "ApiClient") -> str:
+        """
+        Gets a unique id for the current cluster being used.
+
+        There is no real unique identifier for a cluster. However, the `kube-system`
+        namespace is immutable and has a persistence UID that we use instead.
+
+        PREFECT_KUBERNETES_CLUSTER_UID can be set in cases where the `kube-system`
+        namespace cannot be read e.g. when a cluster role cannot be created. If set,
+        this variable will be used and we will not attempt to read the `kube-system`
+        namespace.
+
+        See https://github.com/kubernetes/kubernetes/issues/44954
+        """
+        # Default to an environment variable
+        env_cluster_uid = os.environ.get("PREFECT_KUBERNETES_CLUSTER_UID")
+        if env_cluster_uid:
+            return env_cluster_uid
+
+        # Read the UID from the cluster namespace
+        with self._get_core_client(client) as core_client:
+            namespace = core_client.read_namespace("kube-system")
+        cluster_uid = namespace.metadata.uid
+
+        return cluster_uid
+
+    def _watch_job(
+        self,
+        job_name: str,
+        configuration: KubernetesWorkerJobConfiguration,
+        client: "ApiClient",
+    ) -> int:
+        """
+        Watch a job.
+
+        Return the final status code of the first container.
+        """
+        self._logger.debug(f"Job {job_name!r}: Monitoring job...")
+
+        job = self._get_job(job_name, configuration, client)
+        if not job:
+            return -1
+
+        pod = self._get_job_pod(job_name, configuration, client)
+        if not pod:
+            return -1
+
+        # Calculate the deadline before streaming output
+        deadline = (
+            (time.monotonic() + configuration.job_watch_timeout_seconds)
+            if configuration.job_watch_timeout_seconds is not None
+            else None
+        )
+
+        if configuration.stream_output:
+            with self._get_core_client(client) as core_client:
+                logs = core_client.read_namespaced_pod_log(
+                    pod.metadata.name,
+                    configuration.namespace,
+                    follow=True,
+                    _preload_content=False,
+                    container="prefect-job",
+                )
+                try:
+                    for log in logs.stream():
+                        print(log.decode().rstrip())
+
+                        # Check if we have passed the deadline and should stop streaming
+                        # logs
+                        remaining_time = (
+                            deadline - time.monotonic() if deadline else None
+                        )
+                        if deadline and remaining_time <= 0:
```

Comment:
is `logs.stream()` a blocking call? if so, that means that the deadline is only checked after the blocking call returns, right?

Reply:
From that, my guess is that this means it's a blocking call, so the deadline is only checked when a new log line is printed. This could mean that, if jobs run for a long time but do not print any output, then the deadline may not be checked in that case

Reply:
This is non-trivial to solve so I've opened https://github.com/PrefectHQ/prefect-kubernetes/issues/42 to keep track of this issue.